torch/csrc/jit/subgraph_matcher.cpp

```diff
@@ -0,0 +1,200 @@
+  if (v1->uses().size() != v2->uses().size() && v2->node() != anchor_ &&
+      v1->node()->kind() != prim::Param) {
+    return false;
+  }
+
+  // Add the values to the map before calling matchNodes to avoid infinite
+  // recursion.
+  values_map_[v1] = v2;
+  return matchNodes(v1->node(), v2->node());
+}
+
+/**
+ * Compare two Nodes. N1 is from pattern, N2 is from the actual graph.
+ *
+ * The nodes are considered matching if:
+ * 1) N1 and N2 are of the same kind.
+ * 2) Number of inputs and outputs is the same.
+ * 3) All input and output values match.
+ *
+ * A special case is when N1 is PARAM - this is considered outside the pattern,
+ * so it matches everything.
+ */
+bool SubgraphMatcher::matchNodes(const Node* n1, const Node* n2) {
+  // Check if we've already visited these nodes.
+  if (nodes_map_.count(n1)) {
+    return nodes_map_.at(n1) == n2;
+  }
+
+  // Param node in pattern graph matches everything.
+  if (n1->kind() == prim::Param) {
+    return true;
+  }
+
+  // We don't allow matches to span across blocks, so check if N2 is in the same
+  // block as the first (anchor) node.
+  if (n2->owningBlock() != anchor_->owningBlock()) {
+    return false;
+  }
+
+  if (n1->kind() != n2->kind() ||
+      n1->outputs().size() != n2->outputs().size() ||
+      n1->inputs().size() != n2->inputs().size()) {
+    return false;
+  }
+
+  // Add nodes to the map before calling matchValues to avoid infinite
+  // recursion.
+  nodes_map_[n1] = n2;
+  for (size_t i = 0; i < n1->outputs().size(); i++) {
+    if (!matchValues(n1->outputs()[i], n2->outputs()[i])) {
+      return false;
+    }
+  }
+  for (size_t i = 0; i < n1->inputs().size(); i++) {
+    if (!matchValues(n1->inputs()[i], n2->inputs()[i])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * Recursively try to match pattern with the actual graph starting from the
+ * exiting node in the pattern and anchor node in the actual graph.
+ */
+bool SubgraphMatcher::matchesSubgraphFromAnchorNode(const Node* anchor) {
+  nodes_map_.clear();
+  values_map_.clear();
+  anchor_ = anchor;
+
+  const Node* bottom_node = *(pattern_.nodes().end());
+  AT_ASSERT(bottom_node->inputs().size() == 1);
+  bottom_node = bottom_node->input()->node();
+
+  if (!matchNodes(bottom_node, anchor)) {
+    return false;
+  }
+
+  return true;
+}
+
+} // unnamed namespace
+
+// Main entry point for the subgraph matching.
+std::vector<Match> findPatternMatches(
+    const Graph& pattern,
+    const Graph& graph) {
+  AT_ASSERT(patternGraphIsValid(pattern));
+
+  SubgraphMatcher m(pattern, graph);
+  std::vector<Match> matches;
+  std::stack<const Block*> blocks_to_visit;
+
+  // Iterate over all nodes in the graph (including nodes in subblocks) trying
+  // to match the pattern each node.
+  blocks_to_visit.push(graph.block());
+  while (!blocks_to_visit.empty()) {
+    const Block* block = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (const Node* n : block->nodes()) {
```

Comment:
maybe call out that this is a match on the imperative interpretation of the IR, and not the data flow graph

diamonds declared in different orders won't match, right?

Reply:
Hm, they should match whatever the order is (I added a test for this). We do follow data flow graph, not the sequential list of nodes: `matchNodes` call `matchValues` for nodes' inputs, and `matchValues` proceed to values' defs calling `matchNodes` for them.

PS: This implementation lacks aliasing checks, which are required if we're going to replace matched subgraphs with something else - I don't forget about it, just plan to follow up on that separately.